src/Domain/WeightedElementDistribution.hpp

```diff
@@ -0,0 +1,147 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <array>
+#include <cstddef>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+template <size_t Dim>
+class Block;
+
+template <size_t Dim>
+class ElementId;
+
+namespace Spectral {
+enum class Quadrature;
+}  // namespace Spectral
+
+namespace domain {
+
+/*!
+ * \brief Distribution strategy for assigning elements to CPUs using a
+ * Morton ('Z-order') space-filling curve to determine placement within each
+ * block, where elements are assigned weighted costs that are distributed
+ * across CPUs
+ *
+ * \details The element distribution attempts to assign a balanced total
+ * computational cost to each processor that is allowed to have elements. The
+ * cost assigned to a processor is defined as the sum of the costs of the
+ * `Element`s assigned to a processor, and the cost of an `Element` is defined
+ * as `(number of grid points) / sqrt(minimum grid spacing in Frame::Grid)`.
```

Comment:
Would it be possible with reasonable effort to do the following two things?

- Allow the consumer of this function to pass in the list of all element IDs instead of blocks + refinement levels. This would allow us to periodically re-distribute the elements when we do AMR, as an alternative to Charm's dynamic load balancers (which we're having some trouble with). I expect this will be quite important, and it's probably easier to set up the code to work with element IDs rather than global refinement levels now than to refactor it later.
- Also allow to pass in the weight for each element, so we can experiment with different weightings, instead of hardcoding the min grid spacing. My understanding is that the distribution algorithm is agnostic to the cost function, so decoupling the two would be very helpful. The min grid spacing weighting function can just be a free function that computes the weight given an element ID, mesh, etc.

Reply:
- Yes, the reason I didn't have it accept the list of `ElementId`s as a parameter for the constructor of the distribution is that internally, the weighted distribution needs the elements ordered by z-curve index, so that either requires the caller to correctly call `initial_element_ids_in_z_curve_order` and pass that correct element order into the weighted distribution constructor (which could just assume its input is in the right order, which doesn't seem super safe), *or* we assume the caller would pass in a list of elements in an arbitrary order and internally the weighted distribution would have to sort them by z-curve index to ensure the right ordering is used. In other words, constructing the element list in z-curve order internally guarantees that we use the elements in z-curve order. I think having the constructor or an `update` function take in the list of elements is fine so long as we have a means of ensuring the element ordering used for the distribution is correct (whether that requires we internally sort the input element list or store the elements in z-curve order somewhere and we *know* the input element list is already sorted). Let me know your thoughts on this!
- I agree this is good in principle and the only hangup I have is the same thing as above - that the caller would need to pass in the costs already ordered in z-curve order or the the weighted distribution would internally have to sort the input costs (in which case, costs would need to be paired up with ElementIds).

So based on this, perhaps it seems like one of these things should happen:
- store the ElementIds in z-curve order somewhere so callers can use this and not worry about ordering it correctly
- internally sort the input element lists and costs so callers don't have to worry about ordering it correctly
- anything else that ensures the elements are ordered correctly and we don't have to police each call site to provide the correct order

Reply:
Since, as far as I can tell,  ZCurve assumes all Elements within a Block have the same refinement level in a given dimension, I don't see the point of passing in ElementIds...

Reply:
Oh is that a hard requirement? It would be very nice to be able to load-balance using the weighted z-curve after a few AMR iterations.

Reply:
I haven't looked into space-filling curves, so I don't know how hard it is to implement one that doesn't have that assumption.  But the current `z_curve_index_from_element_id` implementation is clearly not invertible (i.e. multiple ElementIds have the same z_curve_index) if this assumption is broken.

Reply:
You could pass a functor for computing the weight given an element, but I don't know if it would be easy for the functor to actually access anything interesting.

Reply:
After reading through what will be necessary to use this as a basic load balance after AMR, I think it might be best to defer these changes to a later PR. This one is already very long (out of necessity), and I think @geoffrey4444 and I would like to get this in as quick as possible so our BBH runs take less time. Does this seem reasonable @nilsvu @kidder?

p.s. I think the changes suggested here are good and we should definitely do some (all?) of them. Just eventually. Not in this PR.

Reply:
Yes, I agree.

Reply:
A general Z curve does not assume uniform refinement, but our implementation currently does. It will not work with intra-block h-refinement. To me this makes it better to deal with the manual LB in a completely separate PR, because a lot will need to change for that.